shell/platform/tizen/tizen_vsync_waiter.cc

```diff
@@ -16,95 +16,110 @@ namespace flutter {
 
 typedef struct {
   Eina_Thread_Queue_Msg head;
-  int value;
+  int event;
+  intptr_t baton;
 } Msg;
 
-static Eina_Thread_Queue* vblank_thread_queue{nullptr};
-
 TizenVsyncWaiter::TizenVsyncWaiter(FlutterTizenEngine* engine)
     : engine_(engine) {
-  if (!CreateTDMVblank()) {
-    FT_LOGE("Failed to create TDM vblank");
-    DestroyTDMVblank();
-  } else {
-    vblank_thread_queue = eina_thread_queue_new();
-    vblank_thread_ =
-        ecore_thread_feedback_run(RequestVblankLoop, NULL, VblankLoopFinish,
-                                  VblankLoopFinish, this, EINA_TRUE);
-  }
+  vblank_thread_ = ecore_thread_feedback_run(RequestVblankLoop, nullptr,
+                                             nullptr, nullptr, this, EINA_TRUE);
 }
 
 TizenVsyncWaiter::~TizenVsyncWaiter() {
-  SendMessage(kMessageQuit);
+  Send(kMessageQuit, 0);
   if (vblank_thread_) {
     ecore_thread_cancel(vblank_thread_);
     vblank_thread_ = nullptr;
   }
-  DestroyTDMVblank();
 }
 
 void TizenVsyncWaiter::AsyncWaitForVsync(intptr_t baton) {
-  baton_ = baton;
-  if (TDMValid()) {
-    SendMessage(kMessageRequestVblank);
-  }
+  Send(kMessageRequestVblank, baton);
 }
 
-void TizenVsyncWaiter::SendMessage(int val) {
-  if (!vblank_thread_queue || !vblank_thread_) {
-    FT_LOGE("vblank thread or vblank thread queue not valid");
+void TizenVsyncWaiter::OnThreadQueueCreate(
+    Eina_Thread_Queue* vblank_thread_queue) {
+  vblank_thread_queue_ = vblank_thread_queue;
+}
+
+void TizenVsyncWaiter::Send(int event, intptr_t baton) {
+  if (!vblank_thread_ || ecore_thread_check(vblank_thread_)) {
+    FT_LOGE("vblank thread not valid");
+    return;
+  }
+
+  if (!vblank_thread_queue_) {
+    FT_LOGE("vblank thread queue not valid");
     return;
   }
+
   Msg* msg;
   void* ref;
   msg = static_cast<Msg*>(
-      eina_thread_queue_send(vblank_thread_queue, sizeof(Msg), &ref));
-  msg->value = val;
-  eina_thread_queue_send_done(vblank_thread_queue, ref);
+      eina_thread_queue_send(vblank_thread_queue_, sizeof(Msg), &ref));
+  msg->event = event;
+  msg->baton = baton;
+  eina_thread_queue_send_done(vblank_thread_queue_, ref);
 }
 
 void TizenVsyncWaiter::RequestVblankLoop(void* data, Ecore_Thread* thread) {
   TizenVsyncWaiter* tizen_vsync_waiter =
       reinterpret_cast<TizenVsyncWaiter*>(data);
-  void* ref;
-  Msg* msg;
+  TdmClient tdmClient(tizen_vsync_waiter->engine_);
+  if (!tdmClient.IsValid()) {
+    FT_LOGE("Tdm client not valid");
+    ecore_thread_cancel(thread);
+    return;
+  }
+  Eina_Thread_Queue* vblank_thread_queue = eina_thread_queue_new();
+  if (!vblank_thread_queue) {
+    FT_LOGE("Vblank thread queue is not valid");
+    ecore_thread_cancel(thread);
+    return;
+  }
+  tizen_vsync_waiter->OnThreadQueueCreate(vblank_thread_queue);
```

Comment:
I'm just curious. Doesn't the following work in this case?

```cpp
tizen_vsync_waiter->vblank_thread_queue_ = vblank_thread_queue;
```

Reply:
That's ok.